### PR TITLE
feat: shed: Add --below-basefee to mpool clear

### PR DESCRIPTION
**Pull request by magik6k**
## Additional Info
Not tested at all

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/messagepool/messagepool.go [hunk 1]
```diff
@@ -448,6 +448,36 @@ func New(ctx context.Context, api Provider, ds dtypes.MetadataDS, us stmgr.Upgra
...
 
+func (ms *msgSet) removeMessagesBelowBaseFee(baseFee types.BigInt) bool {
+	noncesToRemove := make([]uint64, 0)
+
+	// Find nonces of messages below the base fee and mark the first nonce to remove
+	var firstNonceToRemove uint64 = math.MaxUint64
+	for nonce, msg := range ms.msgs {
+		if types.BigCmp(msg.Message.GasFeeCap, baseFee) < 0 {
+			noncesToRemove = append(noncesToRemove, nonce)
+			if nonce > ms.nextNonce && firstNonceToRemove == math.MaxUint64 {
```
Comment (arajasek) [inline chain/messagepool/messagepool.go:459]: I'm not sure I understand the `nonce > ms.nextNonce` check. Why do we only set `firstNonceToRemove` in this case?

Depending on what the intended logic we want for this, we _probably_ just want to find the lowest nonce that has `GasFeeCap < baseFee`, and then drop all subsequent messages?

File: chain/messagepool/messagepool.go [hunk 2]
```diff
@@ -1584,6 +1614,11 @@ func (mp *MessagePool) Clear(ctx context.Context, local bool) {
 
 			if ok {
 				for _, m := range mset.msgs {
+					if types.BigCmp(co.MinBaseFee, big.Zero()) > 0 &&
+						types.BigCmp(m.Message.GasFeeCap, co.MinBaseFee) < 0 {
```
Comment (arajasek) [inline chain/messagepool/messagepool.go:1618]: We _don't_ delete messages with `GasFeeCap < MinBaseFee`? Because?

**Review by arajasek — COMMENTED** [review]
I'm not sure I understand the core logic here, simple as it is. There's arguably a few different ways we can implement this, and we need to worry about "holes" in our mSets (I think). It looks like you're trying to handle that, but I'm not sure I understand how.

File: chain/messagepool/messagepool.go (hunk 2 above)
Comment (magik6k) [reply chain/messagepool/messagepool.go:1618]: This.. maay be in the wrong order

Comment (magik6k) [reply chain/messagepool/messagepool.go:1618]: That is we should remove messages below the specified minimum basefee

File: chain/messagepool/messagepool.go (hunk 1 above)
Comment (magik6k) [reply chain/messagepool/messagepool.go:459]: That would be the much less dumb way to do that, assuming that message sets are sorted we could just truncate.

Comment (arajasek) [reply chain/messagepool/messagepool.go:459]: They're not, they're a map.

File: chain/messagepool/messagepool.go (hunk 2 above)
Comment (arajasek) [reply chain/messagepool/messagepool.go:1618]: That makes more sense.

**Review by arajasek — COMMENTED** [review]
I'm gonna say let's not merge this unless a pressing need emerges? The mpool really should be doing this kinda thinking itself when necessary.

(Closing for now, but feel free to yell)

**Pull request closed without merging**